### PR TITLE
F/614 deps free umple jar

### PR DESCRIPTION
**Pull request by aforward**
Here is the approach I am taking to pull in jopt-simple for #614 

Note that my editor strips extra spaces at the end of lines, so a little bit of extra diffs.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,9 @@
 bin/
-dist/ 
+dist/
+nohup.out
+
+# Temporary
+umpleonline/index.php
```
Comment (aforward) [inline .gitignore:6]: For "working" in my docker container, I need index.php (our default is umple.php which is then configured through apache / nginx to serve as the default.  I might do a separate PR with both, and then a third to remove it, once we see things working just fine.

File: build/_template.xml [hunk 2]
```diff
@@ -19,14 +19,15 @@
 
   <target name="template.clean">
     <delete dir="cruise.umple.validator/bin" failonerror="false" />
-	<delete dir="cruise.umplificator/bin" failonerror="false" />
+    <delete dir="cruise.umplificator/bin" failonerror="false" />
```
Comment (aforward) [inline build/_template.xml:22]: oh spaces versus tabs :-)

File: build/build.deps.xml [hunk 3]
```diff
@@ -16,6 +16,10 @@
   <property name="umple.deps.manual.dir" value="${umple.deps.dir}/manual" />
   <property name="umple.last.jar.file" value="${umple.deps.dir}/../umple-latest.jar" />
 
+  <!-- Vendor Configuration -->
```
Comment (aforward) [inline build/build.deps.xml:19]: Added the notion of "vendor" which we compile right into umple

File: build/build.umple.xml [hunk 4]
```diff
@@ -40,6 +41,12 @@
     <delete dir="cruise.umple.validator/src-gen-umple" failonerror="false" />
   </target>
 
+  <target name="init">
+    <antcall target="template.init" />
+    <!-- TODO: get the "-f build.deps.xml bootstrap" to be part of the docs / build and then remove this line -->
+    <ant antfile="build/build.deps.xml" target="deps.vendor" inheritAll="true" />
```
Comment (aforward) [inline build/build.umple.xml]: Grabbing vendors doesn't take long, but I would like to make the "bootstrap" more streamlined into our "first-build", but for now I went with the safer always update your vendor dependencies on build

File: build/build.umple.xml [hunk 5]
```diff
@@ -338,9 +346,9 @@
       <classpath refid="core.ivy.classpath"/>
     </manifestclasspath>
 
-    <jar destfile="dist/umple.jar" 
-         basedir="${bin.path}" 
-         includes="cruise/umple/** en.error *.grammar *.error templates/** org/apache/**" 
+    <jar destfile="dist/umple.jar"
+         basedir="${bin.path}"
+         includes="cruise/umple/** en.error *.grammar *.error templates/** org/apache/** joptsimple/**"
```
Comment (aforward) [inline build/build.umple.xml:349]: This is where we shove in joptsimple into our jars.

**Comment by aforward** [comment]
Here's a random directory

```
root@dff360a2844d:/tmp/aha# ls -la
total 10528
drwxr-xr-x 2 root root    4096 Sep 12 17:59 .
drwxrwxrwt 5 root root    4096 Sep 12 17:59 ..
-rw-r--r-- 1 root root      12 Sep 12 13:42 me.ump
-rw-r--r-- 1 root root 5353629 Sep 12 17:59 umple-latest.jar
-rw-r--r-- 1 root root 5409214 Sep 12 17:59 umple.jar
```

Here's what I ran after a build

```
root@dff360a2844d:/tmp/aha# java -jar umple.jar me.ump 
Processing -> me.ump
Success! Processed me.ump.
```

**Comment by aforward** [comment]
The tests are failing, but I think we are further along. Won't be able to look until tomorrow

File: build/_template.xml [hunk 6]
```diff
@@ -19,14 +19,15 @@
 
   <target name="template.clean">
     <delete dir="cruise.umple.validator/bin" failonerror="false" />
-	<delete dir="cruise.umplificator/bin" failonerror="false" />
+    <delete dir="cruise.umplificator/bin" failonerror="false" />
     <delete dir="${bin.path}" failonerror="false" />
     <delete dir="${dist.path}" failonerror="false" />
+    <touch file="dist/.gitkeep"/>
```
Comment (Nava2) [inline build/_template.xml:25]: Good idea on this one.

File: build/build.deps.xml [hunk 7]
```diff
@@ -70,13 +74,27 @@
     </sequential>
   </macrodef>
 
+  <!-- Vendor Dependencies for umple.jar -->
+  <target name="deps.vendor" description="Install vendor dependencies">
```
Comment (Nava2) [inline build/build.deps.xml]: This whole target, could we just expand some of the jars? For example, adding a separate "vendor" configuration within the Ivy file. This means that we can avoid the manual unzipping. 

Example code: 

``` xml
<echo>Installing Vendors...</echo>
<deps-resolve conf="core" pathid="core.ivy.classpath" />
<foreach item="String" in="${core.ivy.classpath}" delim=":" property="jar.path">
  <unjar src="${jar.path}" dest="${some.staging.path.dir}" />
</foreach>

<jar ...>
  <include>${some.staging.path.dir}</include>
</jar>
```

This is very pseudo- but I just don't like the idea of adding manual management of dependencies when they are already downloaded via Ivy.

File: build/build.umple.xml (hunk 4 above)
Comment (Nava2) [inline build/build.umple.xml]: `bootstrap` can only be included _iff_ we remove all references to Ant Contrib in the `first-build` step.

File: build/build.umple.xml [hunk 8]
```diff
@@ -498,7 +506,25 @@
     <antcall target="template.resetVersion" />
   </target>
 
-  <!-- 
+  <!--
+    Run this to simply create a new umple.jar for testing the packaging code
+  -->
+  <target name="package-build">
```
Comment (Nava2) [inline build/build.umple.xml:510]: Good call.

File: build/build.umple.xml [hunk 9]
```diff
@@ -498,7 +506,25 @@
     <antcall target="template.resetVersion" />
   </target>
 
-  <!-- 
+  <!--
+    Run this to simply create a new umple.jar for testing the packaging code
+  -->
+  <target name="package-build">
+    <antcall target="deps:umple" />
```
Comment (Nava2) [inline build/build.umple.xml:511]: `deps.umple` or `deps:umple`, I'm okay with either.

File: build/build.deps.xml (hunk 7 above)
Comment (aforward) [inline build/build.deps.xml]: Yes.  I will keep going down this current line to get the build back up, and then we cal look at leveraging the Ivy to handle the work above

**Comment by aforward** [comment]
The following test is failing in VML, but I cannot seem to find it

```
    [junit] Running cruise.vml.Batch-With-Multiple-Tests
    [junit] Tests run: 1, Failures: 0, Errors: 1, Time elapsed: 0 sec
```

**Comment by aforward** [comment]
It's not VML, it looks like an a possible ANT upgrade

```
 * If a <junit> batch with multiple tests times out Ant logs a message
   about a test named Batch-With-Multiple-Tests since 1.8.0 but the
   logic that determined the Java package of this pseudo-test has been
   wrong.
   Bugzilla Report 45227.
```

https://archive.apache.org/dist/ant/RELEASE-NOTES-apache-ant-1.8.2.html

File: build/build.deps.xml (hunk 7 above)
Comment (Nava2) [inline build/build.deps.xml]: Seems like a good plan.

**Comment by TimLethbridge** [comment]
This pull request seems to work absolutely fine on my Mac. Is it ready for merging into master? Let me know and I will do it (which will result in it then going to cruise control, but I note that a first-build will be required there).

Please note that I am very much a git newbie and am also not as expert as you two are on the build details in this PR.

**Comment by aforward** [comment]
The latest push borked the travis build even more.  I am going to turn off "testing" on travis and we can deal with it separately

**Comment by aforward** [comment]
@TimLethbridge I undid the recent changes to try to get TravisCI to work, still couldn't.

I created two additionals tasks as follow-up

#616 (get travis testing working) and #617 (enable offline building) based on your comments

It would be nice to get TravisCi to someone give a thumbs up, but if this works locally for you, then it might be an ok time to merge it.

**Comment by aforward** [comment]
@TimLethbridge the "build" passes (it doesn't run the tests, just does a first build to ensure things can be packaged --> far from ideal, but we have a ticket open to address that.

**Comment by aforward** [comment]
@TimLethbridge it should be "ok" to push Merge pull request button.  which will grab the code by from fork and branch (aforward:f/614-deps-free-umple-jar) into the core umple project (umple:master).

We have separate tickets setup to continue to get the migration working

**Pull request merged**